clients/imodels-client-authoring/src/operations/changeset/ChangesetOperations.ts

```diff
@@ -261,60 +264,64 @@ export class ChangesetOperations<
       abortSignal: params.abortSignal,
     };
 
-    let bytesDownloaded = 0;
-    if (params.downloadCallback) {
-      downloadParams.latestDownloadedChunkSizeCallback = (downloaded) => {
-        bytesDownloaded += downloaded;
-        params.downloadCallback?.(downloaded);
-      };
-    }
-
-    try {
-      const downloadLink = params.changeset._links.download;
-      assertLink(downloadLink);
-      await downloadFile({
-        ...downloadParams,
-        url: downloadLink.href,
-        storageType: downloadLink.storageType,
-      });
-    } catch (error) {
-      this.throwIfAbortError(error, params.changeset);
-      params.firstDownloadFailedCallback?.(bytesDownloaded);
-
-      const changeset = await this.querySingleInternal({
-        authorization: params.authorization,
-        iModelId: params.iModelId,
-        changesetId: params.changeset.id,
-        headers: params.headers,
-      });
+    let firstError: Error | undefined;
+    for (let i = 0; i < (params.maxRetries ?? 4); i++) {
+      let bytesDownloaded = 0;
+      if (params.downloadCallback) {
+        downloadParams.latestDownloadedChunkSizeCallback = (downloaded) => {
+          bytesDownloaded += downloaded;
+          params.downloadCallback?.(downloaded);
+        };
+      }
 
       try {
-        const newDownloadLink = changeset._links.download;
-        assertLink(newDownloadLink);
+        const downloadLink = await this.getDownloadLink(params, i > 0);
+        assertLink(downloadLink);
         await downloadFile({
           ...downloadParams,
-          url: newDownloadLink.href,
-          storageType: newDownloadLink.storageType,
-        });
-      } catch (errorAfterRetry) {
-        this.throwIfAbortError(error, params.changeset);
-
-        let originalError: Error | undefined;
-        if (errorAfterRetry instanceof Error) originalError = errorAfterRetry;
-
-        throw new IModelsErrorImpl({
-          code: IModelsErrorCode.ChangesetDownloadFailed,
-          message: `Failed to download changeset. Changeset id: ${
-            params.changeset.id
-          }, changeset index: ${
-            params.changeset.index
-          }, error: ${JSON.stringify(errorAfterRetry)}.`,
-          originalError,
-          statusCode: undefined,
-          details: undefined,
+          url: downloadLink.href,
+          storageType: downloadLink.storageType,
         });
+        return;
+      } catch (error) {
+        this.throwIfAbortError(
+          error,
+          params.changeset,
+          downloadParams.abortSignal
+        );
+        if (i === 0) {
+          params.firstDownloadFailedCallback?.(bytesDownloaded);
+        }
+        if (error instanceof Error && firstError == null) firstError = error;
```

Comment:
Not sure if that's possible, but if something else than instance of Error is thrown, this method will silently succeed. Would be good to fail always.